File: asgate/test/session_test.dart
import 'dart:convert';

import 'package:asgate/asgate.dart';
import 'package:test/test.dart';

String _makeJwt(Map<String, dynamic> payload) {
  final header = base64Url.encode(utf8.encode(jsonEncode({'alg': 'none'})));
  final body = base64Url.encode(utf8.encode(jsonEncode(payload)));
  return '${header.replaceAll('=', '')}.${body.replaceAll('=', '')}.sig';
}

void main() {
  group('Session', () {
    final now = DateTime.utc(2026, 8, 17);

    Map<String, dynamic> loginJson() => {
      'access_token': 'at-1',
      'refresh_token': 'rt-1',
      'token_type': 'bearer',
      'expires_in': 3600,
      'expires_at': '2026-08-18T00:00:00.000Z',
      'user': {'id': 'u1', 'email': 'a@b.c', 'full_name': 'A B'},
      'must_change_password': false,
    };

    test('parses a login response', () {
      final session = Session.fromJson(loginJson(), now: now);
      expect(session, isNotNull);
      expect(session!.accessToken, 'at-1');
      expect(session.refreshToken, 'rt-1');
      expect(session.expiresIn, 3600);
      expect(session.user!.email, 'a@b.c');
      expect(session.mustChangePassword, isFalse);
      expect(session.accessTokenExpiresAt, now.add(const Duration(hours: 1)));
      expect(session.sessionExpiresAt, DateTime.utc(2026, 8, 18));
    });

    test('returns null without access_token', () {
      expect(Session.fromJson({'refresh_token': 'rt'}), isNull);
    });

    test('persists and restores absolute access-token expiry', () {
      final session = Session.fromJson(loginJson(), now: now);
      final restored = Session.fromPersisted(session!.persistSessionString);
      expect(restored, isNotNull);
      expect(restored!.accessToken, 'at-1');
      expect(restored.refreshToken, 'rt-1');
      expect(restored.accessTokenExpiresAt, session.accessTokenExpiresAt);
      expect(restored.user!.email, 'a@b.c');
    });

    test('isExpired respects the 30s margin', () {
      final session = Session.fromJson({
        'access_token': 'tok',
        'expires_in': 10,
      }, now: DateTime.now());
      expect(session!.isExpired, isTrue);
    });

    test('isSessionExpired reflects the hard deadline', () {
      final expired = Session(
        accessToken: 'tok',
        accessTokenExpiresAt: DateTime.now().add(const Duration(hours: 1)),
        sessionExpiresAt: DateTime.now().subtract(const Duration(minutes: 1)),
      );
      final alive = Session(
        accessToken: 'tok',
        accessTokenExpiresAt: DateTime.now().add(const Duration(hours: 1)),
        sessionExpiresAt: DateTime.now().add(const Duration(hours: 2)),
      );
      expect(expired.isSessionExpired, isTrue);
      expect(alive.isSessionExpired, isFalse);
    });

    test('fromOidcRedirectUrl parses fragment tokens and JWT claims', () {
      final exp = DateTime.utc(2030).millisecondsSinceEpoch ~/ 1000;
      final jwt = _makeJwt({'sub': 'u1', 'email': 'a@b.c', 'exp': exp});
      final session = Session.fromOidcRedirectUrl(
        'https://site.example.com/auth/callback'
        '#access_token=$jwt&refresh_token=rt-1',
      );
      expect(session, isNotNull);
      expect(session!.accessToken, jwt);
      expect(session.refreshToken, 'rt-1');
      expect(session.user!.id, 'u1');
      expect(session.user!.email, 'a@b.c');
      expect(
        session.accessTokenExpiresAt,
        DateTime.fromMillisecondsSinceEpoch(exp * 1000),
      );
    });

    test('fromOidcRedirectUrl returns null without tokens', () {
      expect(
        Session.fromOidcRedirectUrl(
          'https://site.example.com/auth/callback#x=1',
        ),
        isNull,
      );
    });
  });
}

File: asgate-js/asgate-js/test/session.test.ts
import { describe, expect, it } from 'vitest'
import { Session } from '../src/lib/session'
import { fakeJwt } from './helpers'

describe('Session', () => {
  it('parsea desde JSON con expires_in', () => {
    const session = Session.fromJson(
      {
        access_token: 'tok',
        refresh_token: 'ref',
        token_type: 'bearer',
        expires_in: 3600,
      },
      new Date('2026-01-01T00:00:00Z'),
    )
    expect(session?.accessToken).toBe('tok')
    expect(session?.refreshToken).toBe('ref')
    expect(session?.tokenType).toBe('bearer')
    expect(session?.accessTokenExpiresAt?.toISOString()).toBe(
      '2026-01-01T01:00:00.000Z',
    )
  })

  it('devuelve null sin access_token', () => {
    expect(Session.fromJson({})).toBeNull()
    expect(Session.fromJson({ refresh_token: 'ref' })).toBeNull()
  })

  it('deriva accessTokenExpiresAt del JWT exp', () => {
    const session = Session.fromJson(
      { access_token: fakeJwt({ exp: 2000000000 }) },
      new Date('2026-01-01T00:00:00Z'),
    )
    expect(session?.accessTokenExpiresAt?.toISOString()).toBe(
      '2033-05-18T03:33:20.000Z',
    )
  })

  it('parsea must_change_password y expires_at', () => {
    const session = Session.fromJson({
      access_token: 'tok',
      must_change_password: true,
      expires_at: '2026-01-02T00:00:00Z',
    })
    expect(session?.mustChangePassword).toBe(true)
    expect(session?.sessionExpiresAt?.toISOString()).toBe(
      '2026-01-02T00:00:00.000Z',
    )
  })

  it('round-trip persist/restore', () => {
    const s = Session.fromJson({
      access_token: 'tok',
      refresh_token: 'ref',
      expires_in: 60,
      user: { id: '1', email: 'a@b.c', full_name: 'A' },
    })
    expect(s).not.toBeNull()
    const restored = Session.fromPersisted(s!.persistSessionString)
    expect(restored?.accessToken).toBe('tok')
    expect(restored?.refreshToken).toBe('ref')
    expect(restored?.user?.email).toBe('a@b.c')
  })

  it('fromPersisted devuelve null con JSON inválido', () => {
    expect(Session.fromPersisted('not-json')).toBeNull()
    expect(Session.fromPersisted('{}')).toBeNull()
  })

  it('isExpired aplica margen de 30s', () => {
    const now = Date.now()
    const near = new Session({
      accessToken: 'tok',
      accessTokenExpiresAt: new Date(now + 10_000),
    })
    expect(near.isExpired).toBe(true)
    const far = new Session({
      accessToken: 'tok',
      accessTokenExpiresAt: new Date(now + 60_000),
    })
    expect(far.isExpired).toBe(false)
  })

  it('isSessionExpired usa el vencimiento duro', () => {
    const s = new Session({
      accessToken: 'tok',
      sessionExpiresAt: new Date(Date.now() - 1000),
    })
    expect(s.isSessionExpired).toBe(true)
    const s2 = new Session({
      accessToken: 'tok',
      sessionExpiresAt: new Date(Date.now() + 60_000),
    })
    expect(s2.isSessionExpired).toBe(false)
  })

  it('fromOidcRedirectUrl parsea los tokens del fragment', () => {
    const accessToken = fakeJwt({
      sub: 'u1',
      email: 'a@b.c',
      exp: 2000000000,
    })
    const url = `https://app/auth/callback#access_token=${accessToken}&refresh_token=ref`
    const s = Session.fromOidcRedirectUrl(url)
    expect(s?.accessToken).toBe(accessToken)
    expect(s?.refreshToken).toBe('ref')
    expect(s?.user?.id).toBe('u1')
    expect(s?.user?.email).toBe('a@b.c')
  })

  it('fromOidcRedirectUrl devuelve null sin tokens', () => {
    expect(Session.fromOidcRedirectUrl('https://app/#error=denied')).toBeNull()
  })

  it('cookieSession marca sesión por cookies', () => {
    const s = Session.cookieSession()
    expect(s.isCookieSession).toBe(true)
    expect(s.isExpired).toBe(false)
  })
})
